### PR TITLE
SQL: use a calendar interval for histograms over 1 month intervals

### PR DESCRIPTION
**Pull request by astefan**
Addresses https://github.com/elastic/elasticsearch/issues/51538.

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java [hunk 1]
```diff
@@ -322,17 +320,30 @@ else if (exp instanceof GroupingFunction) {
...
+                                // interval of exactly 1 year or 1 month
+                                if (value instanceof IntervalYearMonth &&
+                                        (((IntervalYearMonth) value).interval().equals(Period.ofYears(1)) 
+                                        || ((IntervalYearMonth) value).interval().equals(Period.ofMonths(1)))) {
+                                    Period yearMonth = ((IntervalYearMonth) value).interval();
+                                    
+                                    String calendarInterval = null;
+                                    if (yearMonth.equals(Period.ofYears(1))) {
+                                        calendarInterval = Histogram.YEAR_INTERVAL;
+                                    } else if (yearMonth.equals(Period.ofMonths(1))) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: This is already checked that it's one of the two so mayb an else is enough.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java [hunk 2]
```diff
@@ -322,17 +320,30 @@ else if (exp instanceof GroupingFunction) {
...
-                                    // When the histogram is `INTERVAL '1' YEAR`, the interval used in the ES date_histogram will be
-                                    // a calendar_interval with value "1y". All other intervals will be fixed_intervals expressed in ms.
-                                    if (field instanceof FieldAttribute) {
-                                        key = new GroupByDateHistogram(aggId, QueryTranslator.nameOf(field), calendarInterval, h.zoneId());
-                                    } else if (field instanceof Function) {
-                                        key = new GroupByDateHistogram(aggId, ((Function) field).asScript(), calendarInterval, h.zoneId());
+                                    // When the histogram is `INTERVAL '1' YEAR` or `INTERVAL '1' MONTH`, the interval used in 
+                                    // the ES date_histogram will be a calendar_interval with value "1y" or "1M" respectively.
+                                    // All other intervals will be fixed_intervals expressed in ms.
+                                    if (calendarInterval != null) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: Here too, the null check is superfluous as the whole block is guarded by: 
```
&& (((IntervalYearMonth) value).interval().equals(Period.ofYears(1)) || ((IntervalYearMonth) value).interval().equals(Period.ofMonths(1)))
```

**Review by matriv — APPROVED** [review]
LGTM. Left 2 minor comments.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java (hunk 2 above)
Comment (astefan) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: Yep, true. It's a leftover from a previous idea.

**Review by bpintea — APPROVED** [review]
LGTM

**Review by costin — APPROVED** [review]
LGTM.

Should we add 1 day as well?

**Comment by astefan** [comment]
> LGTM.
> 
> Should we add 1 day as well?

@costin created https://github.com/elastic/elasticsearch/issues/52713 for `INTERVAL 1 DAY` handling.

**Pull request merged**